Adds Admin Page UI

tellaho opened the pull request:
## Adds admin UI at /admin.

- Ensures that user has custom claim `admin: true;` to access privileged pages
- No "edit" functionality is built-in yet.
- Adds a bunch of shadcn/ui, including those found in `/components/ui` folder

eparubrub committed:
initial admin page data flowing
- adding firebase data converter consume for members

eparubrub committed:
Merge branch 'main' of github.com:hawaiians/hawaiiansintech into eparubrub/admin-page

eparubrub committed:
checkpoint, basic data showing for admin table

eparubrub committed:
pulling updates from eparubrub/firestore-converters

eparubrub committed:
updating yarn.lock

eparubrub committed:
Merge branch 'main' of github.com:hawaiians/hawaiiansintech into eparubrub/admin-page

eparubrub committed:
User removal working

eparubrub committed:
adding email pulling in admin page

eparubrub committed:
minor button adjustment on admin page

tellaho committed:
Merge branch 'main' into eparubrub/admin-page

tellaho committed:
Thank You page + Logo refactor

All because I wanted to use the logo for the admin page! 😆

tellaho committed:
created AdminNav

tellaho committed:
started Dashboard

tellaho committed:
oops removing old Button props

tellaho committed:
moved UserSession to hook

tellaho committed:
LoadingSpinner and Logo tweaks

tellaho committed:
added Emails page

tellaho committed:
more direct (getDoc) method for referencing across tables

since we get reference to em in secureMember collection

tellaho committed:
email list improvements

tellaho committed:
better unsub notice

tellaho committed:
re-added clipboard, sticky nav

tellaho committed:
better layout to inclue transactional notice

tellaho committed:
added more user stories for copying

User stories re: copying emails goes:
1. As an communications admin, I can copy all or only-subscribed emails with a page-, higher-level button.
2. As an communications admin, I can copy individual emails by clicking the card.
3. As an communications admin, I can select multiple emails by clicking on a visible checkbox on the card...
3a. ... then copy those specific email addresses.

tellaho committed:
better email nav

tellaho committed:
improved switcher

tellaho committed:
emails error handling and layout tweaks

tellaho committed:
using radix-ui icons

tellaho committed:
layout clean up

tellaho committed:
text selectable

tellaho committed:
added confirm for >8 selections

tellaho committed:
using new Tabs component

tellaho committed:
added status

tellaho committed:
removed icons (too noisy) and updated confirmation

tellaho committed:
updating homepage to use Tabs too

tellaho committed:
adding subnav to emails

tellaho committed:
better isAdmin gating for AdminNav

tellaho committed:
adding directory route, removing former code

tellaho committed:
removing DECLINED emails, matching style from directory

tellaho committed:
added `Destructive` button variant

tellaho committed:
better permission handling, also fixed nav bug

tellaho committed:
added Modal, very verbose

tellaho committed:
cleaning up unused code

tellaho committed:
trying compound component

tellaho committed:
added lucide icons

tellaho committed:
using Compound components for Admin

learn more, I am really liking this style https://www.youtube.com/watch?v=vPRdY87_SH0

tellaho committed:
using app directory!

tellaho committed:
moved auth logic to admin root layout

tellaho committed:
added passive state, before session check

tellaho committed:
removed old Directory and Emails pages

tellaho committed:
better greeting for admin home

tellaho committed:
inset modal, realizing its time for shadcn

tellaho committed:
added shadcn/ui

tellaho committed:
tweaks

tellaho committed:
started using shadcn Tabs

tellaho committed:
using stub

tellaho committed:
added api route `/admin/members`

tellaho committed:
tried custom user claims (might remove)

tellaho committed:
yeah whack em

tellaho committed:
Revert "removed old Directory and Emails pages"

This reverts commit 8df86bd23dd145de782cd230b2e47cf3b899c106.

tellaho committed:
got old admin working!

tellaho committed:
updated modal

tellaho committed:
oops forgot this one

tellaho committed:
added new tab variants

tellaho committed:
added email lookup

tellaho committed:
fixes for revert gone bad

tellaho committed:
removing old /app structure

tellaho committed:
updated /admin using firebase getAuth()

tellaho committed:
moving isAdmin to hook

tellaho committed:
showing login button regardless of isLoggedIn

tellaho committed:
return all getMembers by default

including StatusEnum only if filtered

tellaho committed:
getMembers now returns regions too

tellaho committed:
better admin checks

tellaho committed:
not sure when nextjs added this but here it is

tellaho committed:
using shadcn Button

tellaho committed:
updating Emails directory

tellaho committed:
passing status as a query param works

tellaho committed:
shifting language for removal to be email (for now)

tellaho committed:
added tooltips to buttons, sorting regions

tellaho committed:
quieted console (server) errors

from this https://github.com/vercel/next.js/issues/14523

tellaho committed:
email list tweaks

tellaho committed:
adjusting use of font-script

tellaho committed:
Merge branch 'main' into tho-admin-page-ui

tellaho committed:
clean up

- removed old components
- forward from /admin to ../directory

tellaho committed:
adds back DECLINED

eparubrub commented on components/Tabs.tsx:
```diff
@@ -0,0 +1,46 @@
+  size?: TabsSize;
+}
+
+export default function Tabs_old({
```
is this function being used anywhere? 

If so what kine naming convention is this? reverse camel_Snake_Case? 😂

tellaho replied:
Currently the homepage still uses these Tabs via `FiltersPicker`, which went unchanged because it imports via an import alias ([this](https://stackoverflow.com/questions/39282253/how-can-i-alias-a-default-import-in-javascript)). I want to replace these but I don't think I'll get to it in this PR; I want to make sure the new Tabs (in `components/ui/tabs`) can be used in both places, probably by styling a new variant.

On leaving it, I thought this weird naming would help us know it's deprecated. But k, I also just used the `@deprecated` JSDoc tag. This makes it so that you'll get a hardy warning in intellisense. Curious if you're cool with this? Or should I just leave them as `Tabs` and the deprecated tag is enough?

<img width="970" alt="image" src="https://github.com/hawaiians/hawaiiansintech/assets/5141111/0ecc2956-55f3-422f-b5a6-b189a69e7646">

eparubrub replied:
gotcha no worries yea if it's still being used then thats all good. makes sense to defer the replacement to another PR. I don't really mind either way in terms of the `@deprecated` tag or not

eparubrub commented on components/ui/checkbox.tsx:
```diff
@@ -0,0 +1,30 @@
+  <CheckboxPrimitive.Root
+    ref={ref}
+    className={cn(
+      "peer h-4 w-4 shrink-0 rounded-sm border border-primary ring-offset-background focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=checked]:text-primary-foreground",
```
Do these not get auto-formatted with our prettier formatter? If not should we find a formatter for these since they go over the line limit most of the time?

tellaho replied:
Yeah, it looks like Prettier doesn't enforce its 80-character default line width for long classnames, like those that Tailwind produces. Found this [stackoverflow](https://stackoverflow.com/questions/75712335/prettier-wont-auto-fix-longer-tailwind-css-class-name-stack-according-to-print) talking about it; it's also discussed here https://github.com/tailwindlabs/tailwindcss/discussions/7763.

[Prettier even tried fixing it then reverted that fix after](https://prettier.io/blog/2021/11/25/2.5.0.html#collapse-html-class-attributes-onto-one-line-11827httpsgithubcomprettierprettierpull11827-by-jlongsterhttpsgithubcomjlongster).

### Solution

**_Note, here's the solution to prevent crazy line widths. I started with /ui components... but I'd prefer to fix the rest in another PR._**

One approach I came upon is to use template literals for long class lists. I don't mind it. What's tough is that it'd be something we'd have to self-enforce (since Prettier won't). Essentially, we'll have to use our judgement when it comes to breaking down longer class lists.

For reference, the class list you commented becomes this:
```
className={cn(
  `peer
  h-4
  w-4
  shrink-0
  rounded-sm
  border
  border-primary
  ring-offset-background
  focus-visible:outline-none
  focus-visible:ring-2
  focus-visible:ring-ring
  focus-visible:ring-offset-2
  disabled:cursor-not-allowed
  disabled:opacity-50
  data-[state=checked]:bg-primary
  data-[state=checked]:text-primary-foreground`,
  className,
)}
```

eparubrub replied:
sounds good, will note to myself to self-enforce which is fine

eparubrub commented on helpers.ts:
```diff
@@ -68,15 +68,22 @@ export function useSessionStorage(key, initialValue) {
+      `@${email.domain}`,
+    ];
+  }
+  return [];
```
would we be showing an empty string in the case that the email is messed up?

tellaho replied:
Ah, so, it looks like the only implementation of `useEmailCloaker()` is when we're adding members via `addMember()`.

And this... doesn't actually make a lot of sense because we immediately convert it to a string in that same addMember function.
```
const maskedEmail = useEmailCloaker(member.email);
const maskedEmailString = `${maskedEmail[0]}...${maskedEmail[1]}${maskedEmail[2]}`;
```

So okay, I made some tweaks to both `useEmailCloaker()` and `addMember()` functions. It now just returns the "masked email string", saving us this roundabout step. But this also necessitated catching errors via a try catch function.

eparubrub commented on lib/api.ts:
```diff
@@ -183,6 +212,73 @@ export async function getMembers(
+        } else {
+          if (!silent)
+            console.warn(
+              `No data available for ${secM.id} in MEMBERS (${secM.fields.email})`
```
curious - were you actually seeing enough of these no data entries to where this was becoming noisy?

tellaho replied:
No longer an issue; removed the `getEmails()` function for the `/api/get-emails` API handler. In which I also cleared this `console.warn()`.

eparubrub commented on pages/admin/directory.tsx:
```diff
@@ -0,0 +1,869 @@
+                  >
+                    <span className="flex items-center gap-2">
+                      <Trash className="h-4 w-4" />
+                      Archive
```
Were you thinking we shouldn't necessarily **Delete** any member data but rather **Disable** them here? 

I think we _should_ be actually removing their data if they're requesting to be removed right? 

Just want to clarify the wording with "Archive"

tellaho replied:
Discussing in Discord.

eparubrub commented on pages/admin/directory.tsx:
A lot of this file seems to be in-progress (such as the delete functionality) but we can come back to it later when I implement write functionality

eparubrub commented on pages/admin/emails.tsx:
```diff
@@ -0,0 +1,497 @@
+    FirebaseTablesEnum.SECURE_MEMBER_DATA
+  );
+
+  const getEmails = secureMemberData.map((secM) => {
```
could we use `getEmails` from the `lib/api.ts` file here?

tellaho replied:
Doing this as a fetch after load, using the new authorization bearer pattern you wrote in the other PR.
```
const fetchEmails = async () => {
  const response = await fetch("/api/get-emails", {
    method: "GET",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${await user.getIdToken()}`,
    },
  });
  const data = await response.json();
  const emails = data.emails.filter((e) => e !== null) as MemberEmail[];
  setEmails(emails);
};
```
Hope you don't mind that I just pasted your latest `auth.tsx` current progress; hoping it's alright we just deal with the conflict if anything changes in your PR.

eparubrub commented on pages/api/get-emails.tsx:
This api endpoint needs to server-side exposed only

tellaho replied:
No longer relevant, doing a token check. Related to https://github.com/hawaiians/hawaiiansintech/pull/23#discussion_r1375366438

eparubrub commented on pages/api/get-members.tsx:
Correct me if I'm wrong but I think this one should be fine exposed since it's basically all the data that's on the HiT page

eparubrub replied:
Actually, I think we don't want to expose this because it will contain data about people who are not approved (pending, in_progress, declined)

tellaho replied:
Did a refactor of `getMembers()` where it does not expose `pending`, `in_progress`, or `declined` members unless you 1) use the fetch function via the handler and 2) include an admin token.

The homepage still uses the normal `getMembers()` function... which does not have access to the token.

eparubrub commented on pages/api/is-admin.tsx:
My gut makes me _feel_ like this should be server side but in a practical sense the public token will be verified and declared whether or not the user is an admin

Is there any way an attacker/hacker could interrupt and mock an api call to return `isAdmin: true`?

eparubrub commented on pages/edit/04-contact.tsx:
```diff
@@ -126,10 +127,26 @@ export default function JoinStep4({ pageTitle }) {
+            kekai@hawaiiansintech.org
+          </Link>{" "}
+          or{" "}
+          <Link href="mailto:kekai@hawaiiansintech.org" target="_blank">
```
```suggestion
          <Link href="mailto:kamakani@hawaiiansintech.org" target="_blank">
```

tellaho replied:
fixed, thanks.

eparubrub reviewed (changes requested):
Most of it is comments/questions. Just need to change a couple of the APIs so that they're only on the server side and we should be good to merge

tellaho committed:
removing unused api endpoint

tellaho committed:
fetching async emails, setting via state

tellaho committed:
setting members via state

tellaho committed:
Revert "removing unused api endpoint"

This reverts commit c51abe392aa833b36b3c203b4abf2c8f11329e62.

tellaho committed:
adds auth.tsx

tellaho committed:
moving email fetch to auth + api endpoint

tellaho committed:
ah caught that I wasn't returning this error

tellaho committed:
Merge branch 'main' into tho-admin-page-ui

tellaho committed:
sync yarn.lock

not sure why but last merge led to conflicts in this yarn-managed file

tellaho committed:
fixed sort order

tellaho committed:
@deprecated warning for Tabs_old

Warns in intellisense

tellaho committed:
fixed email link

tellaho committed:
using /utils cn(), removing helpers.ts cn()

`utils` is preferred place for shad/cn

tellaho committed:
adjusting useEmailCloaker()

- using try catch to handle invalid emails
- this generally shouldn't happen (client-side validation)

tellaho committed:
Using template literal method to shorten /ui components line width

https://github.com/tailwindlabs/tailwindcss/discussions/7763#discussioncomment-4946461

tellaho committed:
moved auth check to getEmails(), from api handler

tellaho committed:
refactor getMembers

eparubrub commented on pages/api/is-admin.tsx:
```diff
@@ -0,0 +1,16 @@
+  await initializeAdmin();
+  if (!token || Object.keys(token).length > 0) {
+    const decodedToken = await admin.auth().verifyIdToken(token);
+    return decodedToken.admin;
```
we could probably just use the `verifyAdminToken` function here right? That function basically has the same logic except it also throws an error for token expiration

eparubrub replied:
could also add the `!token` logic to the `verifyAdminToken` function, maybe we throw an error if no token is provided

Edit*: actually disregard, I can add this in the next PR

tellaho replied:
Nah you know what, I can get to this. Easy.

** oh but I didn't add the `!token` check. Catching that error makes sense.

eparubrub reviewed (approved):
Minah nit, but not a blocker, good to go 🤙

tellaho committed:
using verifyAdminToken for is-admin
